source/logging/StdOutFileLogger.cpp

```diff
@@ -0,0 +1,183 @@
+             << endl;
+        FileUtils::Mkdirs(logFileDir);
+        if (stat(logFileDir.c_str(), &info) != 0)
+        {
+            cout << LOGGER_TAG << ": Failed to create log directories necessary for file-based logging" << endl;
+            return false;
+        }
+        else
+        {
+            cout << LOGGER_TAG << ": Successfully created log directory! Now logging to " << logFile << endl;
+        }
+    }
+    else if (info.st_mode & S_IFDIR)
+    {
+        // Log directory already exists, nothing to do here
+    }
+    else
+    {
+        cout << LOGGER_TAG << ": Unknown condition encountered while trying to create log directory" << endl;
+        return false;
+    }
+
+    // Now we need to establish/verify permissions for the log directory and file
+    if (Permissions::LOG_DIR != FileUtils::GetFilePermissions(logFileDir))
+    {
+        chmod(logFileDir.c_str(), S_IRWXU | S_IRGRP | S_IROTH | S_IXOTH);
+        if (Permissions::LOG_DIR != FileUtils::GetFilePermissions(logFileDir))
+        {
+            cout << LOGGER_TAG
+                 << FormatMessage(
+                        "Failed to set appropriate permissions for log file directory %s, permissions should be set to "
+                        "%d",
+                        logFileDir.c_str(),
+                        Permissions::LOG_DIR);
+        }
+    }
+    outputStream = unique_ptr<ofstream>(new ofstream(logFile, std::fstream::app));
+    if (!outputStream->fail())
+    {
+        if (Permissions::LOG_FILE != FileUtils::GetFilePermissions(logFile))
+        {
+            chmod(logFile.c_str(), S_IRUSR | S_IWUSR);
+            if (Permissions::LOG_FILE != FileUtils::GetFilePermissions(logFile))
+            {
+                cout << LOGGER_TAG
+                     << FormatMessage(
+                            "Failed to set appropriate permissions for log file %s, permissions should be set to %d",
+                            logFile.c_str(),
+                            Permissions::LOG_FILE);
+            }
+        }
+
+        thread log_thread(&StdOutFileLogger::run, this);
+        log_thread.detach();
+        return true;
+    }
+
+    cout << LOGGER_TAG << FormatMessage(": Failed to open %s for logging", logFile.c_str()) << endl;
+    return false;
+}
+
+void StdOutFileLogger::run()
+{
+    unique_lock<mutex> runLock(isRunningLock);
+    isRunning = true;
+    runLock.unlock();
+
+    while (!needsShutdown)
+    {
+        unique_ptr<LogMessage> message = logQueue->getNextLog();
+
+        if (NULL != message)
+        {
+            writeLogMessage(std::move(message));
+        }
+        this_thread::sleep_for(std::chrono::milliseconds(DEFAULT_WAIT_TIME_MILLISECONDS));
+    }
+}
+
+void StdOutFileLogger::stop()
+{
+    needsShutdown = true;
+    logQueue->shutdown();
+
+    unique_lock<mutex> runLock(isRunningLock);
+    isRunning = false;
+}
+
+void StdOutFileLogger::writeLogMessage(unique_ptr<LogMessage> message)
```

Comment:
Except adding logic to write logs to a file and write in `cout` statement, everything written in this file is same as `FileLogger.cpp` and `StdOutLogger.cpp`. Can we instead update the structure of logger feature to avoid this redundancy? instead of creating these three classes, we can only create one single class for `Logger` and add different methods to write logs in file and/or Standard output.